src/vivarium/framework/results/stratification.py

```diff
@@ -0,0 +1,66 @@
+from dataclasses import dataclass
+from typing import Callable, List, Union
+
+import pandas as pd
+from pandas.api.types import CategoricalDtype
+
+
+@dataclass
+class Stratification:
+    """Class for stratifying observed quantities by specified characteristics
+
+    Each Stratification represents a set of mutually exclusive and collectively
+    exhaustive categories into which simulants can be assigned.
+
+    The `Stratification` class has five fields: `name`, `source`, `mapper`,
+    `categories`, and `is_vectorized`. The `name` is the name of the column
+    created by the mapper. The `source` is a list of columns in the extended
+    state table that are the inputs to the mapper function.  Simulants will
+    later be grouped by this column (or these columns) during stratification.
+    `categories` is a set of values that the mapper is allowed to output. The
+    `mapper` is the method that transforms the source to the name column.
+    The method produces an output column by calling the mapper on the source
+    columns. If the mapper is `None`, the default identity mapper is used. If
+    the mapper is not vectorized this is performed by using `pd.apply`.
+    Finally, `is_vectorized` is a boolean parameter that signifies whether
+    mapper function is applied to a single simulant (`False`) or to the whole
+    population (`True`).
+
+    `Stratification` also has a `__call__()` method. The method produces an
+    output column by calling the mapper on the source columns.
+    """
+
+    name: str
+    sources: List[str]
+    categories: List[str]
+    mapper: Callable[[Union[pd.Series, pd.DataFrame]], Union[str, pd.Series]] = None
+    is_vectorized: bool = False
+
+    def __post_init__(self):
+        if self.mapper is None:
+            if len(self.sources) != 1:
+                raise ValueError(
+                    f"No mapper provided for stratification {self.name} with"
+                    f"{len(self.sources)} stratification sources."
+                )
+            self.mapper = self._default_mapper
+            self.is_vectorized = True
+
+    def __call__(self, population: pd.DataFrame) -> pd.DataFrame:
```

Comment:
Should variable name be `population` or would somethign like `source_data` be better?

Reply:
We definitely use `population` in a lot of places in vivarium and refer to discrete elements of that population as "simulants". I think we should keep `population` for consistency. After having done a refactor to `source_data`, you've then got a problem of describing what the `source_data` is, and generally, that's a population with simulants (humans, maybe, but could be boids or something else).